File: LeetCode/0160. Intersection of Two Linked Lists/Solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

 /**
 * 39 / 39 test cases passed.
 * Runtime: 48 ms
 * Memory Usage: 14.2 MB 
 */

class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        ListNode* pA = headA;
        ListNode* pB = headB; 
        bool flagA = false, flagB = false;
        while (pA != pB) {
            if (pA->next) pA = pA->next;
            else if (flagA) break;
            else {
                pA = headB;
                flagA = true;
            }

            if (pB->next) pB = pB->next;
            else if (flagB) break;
            else {
                pB = headA;
                flagB = true;
            }
        }
        return pA == pB ? pA : nullptr;
    }
};

 /**
 * 39 / 39 test cases passed.
 * Runtime: 48 ms
 * Memory Usage: 14.1 MB 
 */
class Solution2 {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        ListNode* pA = headA, *pB = headB;  
        while (pA != pB) {
            pA = pA ? pA->next : headB;
            pB = pB ? pB->next : headA;
        }
        return pA;
    }
};

File: LeetCode/0160. Intersection of Two Linked Lists/solution.py

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

"""
39 / 39 test cases passed.
Runtime: 152 ms
Memory Usage: 29.6 MB
"""
class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> ListNode:
        A, B = headA, headB
        while A != B:
            A = A.next if A else headB
            B = B.next if B else headA
        return A
